sim/core/direct_cast.go

```diff
@@ -174,6 +175,74 @@ func (spell *SimpleSpell) GetDuration() time.Duration {
 	}
 }
 
+func (spell *SimpleSpell) CastAuto(sim *Simulation) {
+	cast := &spell.SpellCast
+	effect := &spell.Effect
+	character := spell.Character
+
+	character.OnBeforeSpellHit(sim, cast, effect)
+	effect.Target.OnBeforeSpellHit(sim, cast, effect)
+
+	effect.Outcome = effect.WhiteHitTableResult(sim, spell)
+
+	if effect.Landed() {
+		bonusWeaponDamage := character.PseudoStats.BonusDamage + effect.BonusWeaponDamage
+
+		dmg := 0.0
+		if !effect.WeaponInput.Offhand {
+			attackPower := character.stats[stats.AttackPower] + effect.BonusAttackPower + effect.BonusAttackPowerOnTarget
+			dmg += character.AutoAttacks.MH.calculateWeaponDamage(sim, attackPower) + bonusWeaponDamage
+		} else {
+			attackPower := character.stats[stats.AttackPower] + effect.BonusAttackPower + 2*effect.BonusAttackPowerOnTarget
+			dmg += character.AutoAttacks.OH.calculateWeaponDamage(sim, attackPower)*0.5 + bonusWeaponDamage
+		}
+
+		dmg += effect.WeaponInput.FlatDamageBonus
+		dmg *= effect.WeaponInput.DamageMultiplier
+
+		dmg += effect.DirectInput.FlatDamageBonus
+
+		if effect.Outcome == OutcomeCrit {
+			dmg *= spell.CritMultiplier
+		} else if effect.Outcome == OutcomeGlance {
+			// TODO glancing blow damage reduction is actually a range ([65%, 85%] vs. 73)
+			dmg *= 0.75
+		}
+
+		// Apply damage reduction.
+		dmg *= 1 - effect.Target.ArmorDamageReduction(character.stats[stats.ArmorPenetration]+effect.BonusArmorPenetration)
+
+		// Apply all other effect multipliers.
+		dmg *= effect.DamageMultiplier * effect.StaticDamageMultiplier
+
+		effect.Damage += dmg
+	}
```

Comment:
Yes, of course. This is not how it's supposed to be done, it's just to illustrate how small the actual cast code is.

Reply:
Ah ok. Yeah I agree there's overhead in the current strategy and welcome any ideas for improving performance. The key is that the the interface to the core modules needs to be as simple as possible, so that it remains easy (and ideally gets easier) for newer programmers to contribute. Simple interfaces also reduce errors.